v2/inc/c_logging/log_context.h

```diff
@@ -4,6 +4,241 @@
+
+// These is an internal API and it is not meant to be called by the users of this module
+int internal_log_context_init_from_parent(LOG_CONTEXT_HANDLE dest_log_context, LOG_CONTEXT_HANDLE parent_log_context);
+uint32_t internal_log_context_get_values_data_length_or_zero(LOG_CONTEXT_HANDLE log_context);
+
+static uint32_t internal_log_context_get_property_value_pair_count_or_zero(LOG_CONTEXT_HANDLE log_context)
+{
+    uint32_t result = log_context_get_property_value_pair_count(log_context);
+    if (result == UINT32_MAX)
+    {
+        result = 0;
+    }
+    return result;
+}
+
+// macro set used to define a parameter in a function signature in order
+// to make sure that no properties with the same name are added in one context
+
+// DEFINE_PROPERTY_AS_PARAMETER
+
+#define EXPAND_DEFINE_PROPERTY_AS_PARAMETER_LOG_CONTEXT_STRING_PROPERTY(property_name, ...) \
+    , int property_name
+
+#define EXPAND_DEFINE_PROPERTY_AS_PARAMETER_LOG_CONTEXT_NAME(log_context_name) \
+
+#define EXPAND_DEFINE_PROPERTY_AS_PARAMETER_LOG_CONTEXT_PROPERTY(property_type, property_name, field_value) \
+    , int property_name
+
+#define DEFINE_PROPERTY_AS_PARAMETER(field_desc) \
+    MU_C2(EXPAND_DEFINE_PROPERTY_AS_PARAMETER_, field_desc)
+
+// macro set used to define a parameter in a function signature  every time LOG_CONTEXT_NAME is used.
+// if used twice a field with the same name is defined => compiler error
+
+// DEFINE_LOG_CONTEXT_NAME_AS_FIELD
+
+#define EXPAND_DEFINE_CONTEXT_NAME_AS_PARAMETER_LOG_CONTEXT_STRING_PROPERTY(property_name, ...) \
+
+#define EXPAND_DEFINE_CONTEXT_NAME_AS_PARAMETER_LOG_CONTEXT_NAME(log_context_name) \
+    , int log_context_is_used_multiple_times
+
+#define EXPAND_DEFINE_CONTEXT_NAME_AS_PARAMETER_LOG_CONTEXT_PROPERTY(property_type, property_name, field_value) \
+
+#define DEFINE_CONTEXT_NAME_AS_PARAMETER(field_desc) \
+    MU_C2(EXPAND_DEFINE_CONTEXT_NAME_AS_PARAMETER_, field_desc)
+
+#define LOG_CONTEXT_CHECK_VARIABLE_ARGS(...) \
+    /* Codes_SRS_LOG_CONTEXT_01_019: [ If 2 properties have the same name for a context a compiler error shall be emitted. ]*/ \
+    (void)(void (*)(int MU_FOR_EACH_1(DEFINE_PROPERTY_AS_PARAMETER, __VA_ARGS__)))0x4242; \
+    /* Codes_SRS_LOG_CONTEXT_01_011: [ If LOG_CONTEXT_NAME is specified multiple times a compiler error shall be emitted. ]*/ \
+    (void)(void (*)(int MU_FOR_EACH_1(DEFINE_CONTEXT_NAME_AS_PARAMETER, __VA_ARGS__)))0x4242; \
+
+// SETUP_PROPERTY_PAIR
+
+#define EXPAND_SETUP_PROPERTY_PAIR_LOG_CONTEXT_STRING_PROPERTY(property_name, ...) \
+    /* Codes_SRS_LOG_CONTEXT_01_007: [ LOG_CONTEXT_STRING_PROPERTY shall expand to code allocating a property/value pair of type ascii_char_ptr and the name property_name. ]*/ \
+    property_value_pair->value = data_pos; \
+    property_value_pair->name = MU_TOSTRING(property_name); \
+    property_value_pair->type = &ascii_char_ptr##_log_context_property_type; \
```

Comment:
>&ascii_char_ptr##_log_context_property_type [](http://example.com/codeflow?start=32&length=43)

sigh. CAn this be written without ##?

Reply:
Kind of hard, so I went for the easy way out